### PR TITLE
Add datatype support for z3

### PR DESCRIPTION
**Pull request by farmerzhang1**
Hi all,

I'm trying to play with smt-switch and decide to add z3 backend support for datatypes. Any feedback is welcome.

### CODE REVIEW

File: contrib/setup-z3.sh [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
 mkdir -p $DEPS
 
+if [ "$(uname)" == "Darwin" ]; then
+    NUM_CORES=$(sysctl -n hw.logicalcpu)
+elif [ "$(expr substr $(uname -s) 1 5)" == "Linux" ]; then
+    NUM_CORES=$(nproc)
+else
+    NUM_CORES=1
+fi
+
```
Comment (yoni206) [inline contrib/setup-z3.sh:18]: I am not sure about this, perhaps this can go to a separate PR, while giving the user the option to choose number of cores?

File: contrib/setup-z3.sh [hunk 2]
```diff
@@ -18,10 +26,20 @@ if [ ! -d "$DEPS/z3" ]; then
     # a pthread related issue
     # compiling with --single-threaded helps, but isn't a real solution
     # see https://github.com/Z3Prover/z3/issues/4554
-    ./configure --staticlib --single-threaded
-    cd build
-    make -j$(nproc)
+    # ./configure --staticlib --single-threaded
+    cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DZ3_BUILD_LIBZ3_SHARED=Off -DZ3_BUILD_LIBZ3_MSVC_STATIC=On
+    cmake --build build -j$NUM_CORES
```
Comment (yoni206) [inline contrib/setup-z3.sh:31]: Is this change necessary? If so, I would delete the commented out line.

File: contrib/setup-z3.sh [hunk 3]
```diff
@@ -18,10 +26,20 @@ if [ ! -d "$DEPS/z3" ]; then
...
+
+if [ -f $DEPS/z3/build/libz3.a ]; then
+    echo "It appears z3 was setup successfully into $DEPS/z3."
+    echo "You may now install it with make ./configure.sh --z3 && cd build && make"
+else
+    echo "Building z3 failed."
+    echo "You might be missing some dependencies."
+    echo "Please see their github page for installation instructions: https://github.com/z3/z3"
+    exit 1
+fi
```
Comment (yoni206) [inline contrib/setup-z3.sh:45]: same question -- is this necessary? What is the current behavior?

File: z3/include/z3_datatype.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
+#pragma once
+#include <memory>
+#include <string>
+#include <vector>
```
Comment (yoni206) [inline z3/include/z3_datatype.h:4]: Are these needed? I don't see them in cvc5_datatype.h.

File: z3/include/z3_datatype.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+#include "exceptions.h"
+#include "z3++.h"
+
+namespace smt {
+
+// forward declaration
+class Z3Solver;
+class Z3Datatype;
+class Z3DatatypeConstructorDecl;
+class Z3DatatypeDecl : public AbsDatatypeDecl
```
Comment (yoni206) [inline z3/include/z3_datatype.h:16]: Same question -- I don't see forward declarations in the cvc5 case. Can we try to make the z3 and cvc5 cases uniform? Or is that really needed only for z3?

File: z3/include/z3_datatype.h [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+{
+ public:
+  Z3DatatypeDecl(z3::context & c, std::string name) : c(c), name(name){};
+
+ protected:
+  friend class Z3Solver;
+  friend class Z3Datatype;
```
Comment (yoni206) [inline z3/include/z3_datatype.h:23]: same

File: z3/src/z3_datatype.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
+#include "z3_datatype.h"
```
Comment (yoni206) [inline z3/src/z3_datatype.cpp:1]: This file is also pretty different from the cvc5 version, which is way shorter. Is that needed?

**Review by yoni206 — COMMENTED** [review]
Thank you very much for this PR. It would be great to have support for z3 datatypes!
I've made some initial comments/questions inline.

In addition to them -- I wonder -- would it make sense to add tests that are specific to z3 datatypes? Perhaps here?
https://github.com/stanford-centaur/smt-switch/tree/main/tests/z3

We are already in good shape because I can see that CI ran the general datatypes tests, but I still wonder whether there is anything specific to test in the z3 API side.

File: contrib/setup-z3.sh (hunk 1 above)
Comment (farmerzhang1) [reply contrib/setup-z3.sh:18]: Well, I am copying this from contrib/setup-cvc5.sh

File: contrib/setup-z3.sh (hunk 2 above)
Comment (farmerzhang1) [reply contrib/setup-z3.sh:31]: Sure, I'll remove the comment

File: contrib/setup-z3.sh (hunk 3 above)
Comment (farmerzhang1) [reply contrib/setup-z3.sh:44]: Copied from contrib/setup-cvc5.sh as well.

File: z3/include/z3_datatype.h (hunk 4 above)
Comment (farmerzhang1) [reply z3/include/z3_datatype.h]: Removed

File: z3/include/z3_datatype.h (hunk 5 above)
Comment (farmerzhang1) [reply z3/include/z3_datatype.h]: Removed

File: z3/src/z3_datatype.cpp (hunk 7 above)
Comment (farmerzhang1) [reply z3/src/z3_datatype.cpp:1]: Done. I kept only the compare function and the auxiliary functions for it.

File: z3/include/z3_datatype.h (hunk 6 above)
Comment (farmerzhang1) [reply z3/include/z3_datatype.h]: Removed

**Comment by farmerzhang1** [comment]
Hi Yoni, I've add some z3-specific tests and try to mirror the cvc5 datatype files as much as possible. Maybe you can check them again?

File: contrib/setup-z3.sh [hunk 8]
```diff
@@ -18,10 +26,19 @@ if [ ! -d "$DEPS/z3" ]; then
     # a pthread related issue
     # compiling with --single-threaded helps, but isn't a real solution
     # see https://github.com/Z3Prover/z3/issues/4554
-    ./configure --staticlib --single-threaded
-    cd build
-    make -j$(nproc)
+    cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DZ3_BUILD_LIBZ3_SHARED=Off -DZ3_BUILD_LIBZ3_MSVC_STATIC=On
+    cmake --build build -j$NUM_CORES
```
Comment (yoni206) [inline contrib/setup-z3.sh:30]: Why is this needed?

**Review by yoni206 — COMMENTED** [review]
Thanks for the changes. Looks like the new tests are passing, and also the old ones, also after adding datatypes to the list of supported theories of z3, which is great.

Just one last question inline.

**Review by yoni206 — APPROVED** [review]
Thanks again for this valuable contribution!
